release-19.2: c-deps/rocksdb: fix correctness issue with snapshots and delete range

petermattis opened the pull request:
Fix a bug where RocksDB allowed a range deletion added after a snapshot
was created to incorrectly affect reads from the snapshot if the range
deletion was present in an immutable memtable.

Release note (bug fix): Fix a bug that could lead to data corruption or
data loss if a replica was both the source of a snapshot and was being
concurrently removed from the range and certain specific conditions
exist inside RocksDB. This scenario is rare, but possible.

petermattis committed:
release-19.2: c-deps/rocksdb: fix correctness issue with snapshots and delete range

Fix a bug where RocksDB allowed a range deletion added after a snapshot
was created to incorrectly affect reads from the snapshot if the range
deletion was present in an immutable memtable.

Release note (bug fix): Fix a bug that could lead to data corruption or
data loss if a replica was both the source of a snapshot and was being
concurrently removed from the range and certain specific conditions
exist inside RocksDB. This scenario is rare, but possible.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48319)
<!-- Reviewable:end -->

petermattis committed:
c-deps: bump the rebuild counter for all cmake dependencies

This is a manual backport of #46796. It makes all of the c-deps rebuild
counters different values for 19.2 vs master.

We recently bumped the version of cmake in the builder. This caused
strange build errors on CI due to switching back and forth between the
new version of cmake and the old version cmake. The makefiles created by
cmake seem to depend on the installed version of cmake in a way I don't
quite understand.

Rather than trying to add a dependency on the version of cmake, simply
bump the rebuild counter for all cmake dependencies. The one exception
is libgeos which requires the new version of cmake (and landed after the
cmake version bump was performed).

Fixes cockroachdb/dev-inf/issues/66

Release note: None

petermattis commented:
```
[Compile C dependencies] make[1]: *** [cmake_check_build_system] Error 1
[14:46:56][Compile C dependencies] Makefile:4299: recipe for target 'cmake_check_build_system' failed
[14:46:56][Compile C dependencies] Makefile:419: recipe for target 'cmake_check_build_system' failed
[14:46:56][Compile C dependencies] CMake Error: The source directory "" does not exist.
```

I've added another commit to this PR which gives the c-deps "rebuild counters" unique values on the 19.2 branch. This is essentially a manual backport of #46796.

petermattis commented:
TFTR!
